cli_tools/gce_windows_upgrade/upgrader/utils.go

```diff
@@ -0,0 +1,147 @@
+		"1. Detach the new OS disk from the instance and delete it.\n" +
+		"2. Attach the old OS disk as boot disk.\n" +
+		"3. Detach the install media disk from the instance and delete it.\n" +
+		"4. Delete 'windows-startup-script-url' from the instance's metadata if there isn't an original value. " +
+		"If there is an original value, restore it. The original value is backed up as metadata 'windows-startup-script-url-backup'.\n" +
+		"\n" +
+		"Once you verified the upgrading succeeded and decided to never rollback, you can:\n" +
+		"1. Delete the original OS disk.\n" +
+		"2. Delete the machine image.\n" +
+		"3. Delete the snapshot.\n" +
+		"\n"
+)
+
+func reverseMap(m map[string]string) map[string]string {
+	newMap := make(map[string]string, len(m))
+	for k, v := range m {
+		newMap[v] = k
+	}
+	return newMap
+}
+
+func getKeys(m map[string]string) []string {
+	keys := make([]string, 0, len(m))
+	for k := range m {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+// SupportedSourceOSVersions returns supported source versions of upgrading
+func SupportedSourceOSVersions() []string {
+	return getKeys(supportedTargetOSVersions)
+}
+
+// SupportedTargetOSVersions returns supported target versions of upgrading
+func SupportedTargetOSVersions() []string {
+	return getKeys(supportedTargetOSVersions)
+}
+
+func getUpgradeGuide(u *Upgrader) (string, error) {
+	originalStartupScriptURL := "None."
+	if u.windowsStartupScriptURLBackup != nil {
+		originalStartupScriptURL = *u.windowsStartupScriptURLBackup
+	}
+	if u.machineImageBackupName == "" {
+		u.machineImageBackupName = "Not created. Machine Image backup is disabled."
+	}
+
+	t, err := template.New("guide").Option("missingkey=error").Parse(upgradeIntroductionTemplate)
+	if err != nil {
+		return "", daisy.Errf("Failed to parse upgrade guide.")
+	}
+	var buf bytes.Buffer
+	varMap := map[string]interface{}{
+		"project":                  u.project,
+		"zone":                     u.zone,
+		"instanceName":             u.instanceName,
+		"installMediaDiskName":     u.installMediaDiskName,
+		"osDiskSnapshotName":       u.osDiskSnapshotName,
+		"osDiskName":               daisyutils.GetResourceRealName(u.osDiskURI),
+		"osDiskDeviceName":         u.osDiskDeviceName,
+		"osDiskAutoDelete":         u.osDiskAutoDelete,
+		"newOSDiskName":            u.newOSDiskName,
+		"machineImageName":         u.machineImageBackupName,
+		"originalStartupScriptURL": originalStartupScriptURL,
+	}
+	if err := t.Execute(&buf, varMap); err != nil {
+		return "", daisy.Errf("Failed to generate upgrade guide.")
+	}
+	return string(buf.Bytes()), nil
+}
+
+func isNewOSDiskAttached(project, zone, instanceName, newOSDiskName string) bool {
+	inst, err := computeClient.GetInstance(project, zone, instanceName)
+	if err != nil {
+		// failed to fetch info. Can't guarantee new OS disk is attached.
+		return false
+	}
+	if inst.Disks == nil || len(inst.Disks) == 0 || inst.Disks[0].Boot == false {
```

Comment:
Can you add some comments:

1. In the 'normal' case, when would you find an instance without disks?
2. Add link to documentation that shows the first disk is always the boot disk.

Reply:
added